packages/ckeditor5-bookmark/src/insertbookmarkcommand.ts

```diff
@@ -0,0 +1,127 @@
+	}
+
+	/**
+	 * Executes the command.
+	 *
+	 * @fires execute
+	 * @param options Command options.
+	 * @param options.bookmarkId The value of the `bookmarkId` attribute.
+	 */
+	public override execute( options: { bookmarkId: string } ): void {
+		if ( !options ) {
+			return;
+		}
+
+		const { bookmarkId } = options;
+
+		if ( !bookmarkId || typeof bookmarkId !== 'string' ) {
+			return;
+		}
+
+		const editor = this.editor;
+		const model = editor.model;
+		const selection = model.document.selection;
+
+		model.change( writer => {
+			const startPosition = selection.isCollapsed ? selection.anchor! : selection.getFirstPosition()!;
+			const isBookmarkAllowedBySchema = model.schema.checkChild( startPosition, 'bookmark' );
+
+			// If it is allowed to insert bookmark by schema than insert it.
+			if ( isBookmarkAllowedBySchema ) {
+				writer.setSelection( startPosition );
+
+				return model.insertObject( writer.createElement( 'bookmark', { bookmarkId } ) );
+			} else {
+				const selectedElement = selection.getSelectedElement();
+
+				// If is selected single element and if it's a block element
+				// check if paragraph can be inserted before it and add bookmark inside of this paragraph.
+				if ( selectedElement && model.schema.isBlock( selectedElement ) ) {
+					editor.execute( 'insertParagraph', {
+						position: editor.model.createPositionAt( selectedElement, 'before' )
+					} );
+
+					return model.insertObject( writer.createElement( 'bookmark', { bookmarkId } ) );
+				} else {
+					// If paragraph is not allowed by the schema then scan the selection ranges to find the first position
+					// for bookmark insertion (for example multiple table cells selected)
+					const allowedElement = this._getAllowedElementBasedOnSelection( selection ) as Element | null;
+
+					if ( allowedElement && allowedElement.childCount ) {
+						const firstChild = allowedElement.getChild( 0 )!;
+
+						// When first child is a paragraph then insert bookmark at the first place inside of it.
+						if ( firstChild.is( 'element', 'paragraph' ) ) {
+							writer.setSelection( firstChild, 'before' );
+
+							return writer.insertElement( 'bookmark', { bookmarkId }, firstChild );
+						}
+
+						// If first child is not a paragraph then insert paragraph before it and add bookmark inside of this paragraph.
+						editor.execute( 'insertParagraph', {
+							position: model.createPositionAt( allowedElement, 0 )
+						} );
+
+						return model.insertObject( writer.createElement( 'bookmark', { bookmarkId } ) );
+					}
+				}
+			}
+		} );
+	}
+
+	/**
+	 * Returns the first element that accepts adding a `paragraph` on it.
+	 * @param selection Current selection.
+	 */
+	private _getAllowedElementBasedOnSelection( selection: Selection | DocumentSelection ): Item | null {
+		const schema = this.editor.model.schema;
+
+		for ( const range of selection.getRanges() ) {
```

Comment:
Those ranges are not sorted, we need to find the first range `selection.getFirstRange()`.